### PR TITLE
inject a provenance subdocument to device data

### PR DESCRIPTION
**Pull request by ewollesen**
BACK-2504

### CODE REVIEW

File: data/service/api/v1/datasets_data_create.go [hunk 1]
```diff
@@ -95,10 +94,18 @@ func DataSetsDataCreate(dataServiceContext dataService.Context) {
 		return
 	}
 
+	var provenance *data.Provenance
+	if claims, err := dataServiceContext.AuthClient().ServerSessionTokenClaims(ctx); err != nil {
```
Comment (toddkazakov) [inline data/service/api/v1/datasets_data_create.go:98]: This doesn't look correct - we should get the client of the access token that was passed in the `x-tidepool-session-token` header.

Comment (ewollesen) [reply data/service/api/v1/datasets_data_create.go]: Yeah, you caught this for review before it was ready, sorry... But all the same, you're right, that's the wrong token from which to grab the client id. 😄 

I'll move it to a draft.

File: Makefile [hunk 2]
```diff
@@ -112,7 +112,7 @@ format-write-changed:
 imports: goimports
 	@echo "goimports -d -e -local 'github.com/tidepool-org/platform'"
 	@cd $(ROOT_DIRECTORY) && \
-		O=`find . -not -path './vendor/*' -name '*.go' -type f -exec goimports -d -e -local 'github.com/tidepool-org/platform' {} \; 2>&1` && \
+		O=`find . -not -path './vendor/*' -not -path '**/test/mock.go' -name '*.go' -type f -exec goimports -d -e -local 'github.com/tidepool-org/platform' {} \; 2>&1` && \
```
Comment (ewollesen) [inline Makefile:115]: To add a little detail, not only are they generated in a way that goimports doesn't like, but then Travis fails the build as a result.

File: data/service/api/v1/datasets_data_create_provenance_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
+package v1
```
Comment (toddkazakov) [inline data/service/api/v1/datasets_data_create_provenance_test.go:1]: Package should be `v1_test` for consistency

**Review by toddkazakov — CHANGES_REQUESTED** [review]
Looks good overall. I believe we need to set the provenance to data sets (upload records) as well.

File: data/service/api/v1/datasets_data_create_provenance_test.go (hunk 3 above)
Comment (ewollesen) [reply data/service/api/v1/datasets_data_create_provenance_test.go:1]: If I do that, then I can't call `selectXFF` or 'collectProvenanceInfo`, as they're not exported, and there's no reason for them to be.

I'm all for testing the exported APIs, but personally I have no problem with going beyond and testing non-exported functions too.

I suppose I could scaffold a bunch of junk to test selectXFF or collectProvenanceInfo via the exported API, but what does that get me? My selectXFF tests get a lot more complicated and brittle, and if selectXFF changes, I'll still have to fix all the tests.

So I'd rather keep the package v1, but if consistency is more important, reply back and I'll do the scaffolding route... Unless you have a better idea?

File: data/service/api/v1/datasets_data_create.go [hunk 4]
```diff
@@ -124,3 +132,83 @@ func DataSetsDataCreate(dataServiceContext dataService.Context) {
...
+
+// collectProvenanceInfo from a request and its auth details.
+//
+// All work is done as a best effort right now.
+func collectProvenanceInfo(ctx context.Context, req *rest.Request, details request.AuthDetails) *data.Provenance {
+	lgr := log.LoggerFromContext(ctx)
+	provenance := &data.Provenance{}
+
+	token := ""
+	if _, ok := req.Header[auth.TidepoolSessionTokenHeaderKey]; ok {
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply data/service/api/v1/datasets_data_create.go]: Fixed

**Comment by ewollesen** [comment]
> Looks good overall. I believe we need to set the provenance to data sets (upload records) as well.

OK, will do.

File: data/service/api/v1/datasets_data_create_provenance_test.go (hunk 3 above)
Comment (toddkazakov) [reply data/service/api/v1/datasets_data_create_provenance_test.go:1]: This exact same issue tripped me [here](https://github.com/tidepool-org/clinic/commit/5e1381a214be06dc12963906c2dcbef3ab82f22f). I don't see any harm in exporting those functions.

**Review by toddkazakov — APPROVED** [review]

**Pull request merged**